### PR TITLE
Create endpoints For CRUD Announcements

### LINKED ISSUE #7: Create endpoints to CRUD News and Announcements
- The data model can be found in the Google doc.

### PR DESCRIPTION
**Pull request by atakoutene**
## WHY
- Closes #7

## WHAT
1- Design logic to handle Announcement, cancel News (didn't find it relevant for this scope, let's treat everything as announcement for the beginning)
2- An announcement is uniquely identified by its title, but I've added a surrogate key id (auto increment)
3- I didn't find it useful to add the author of the announcement, as everything is posted by the community executive from the back office! (What is your thought?)
4- Handled all endpoints for CRUD.
5- Designed Unit tests for everything.
## TESTING

test-community001

### CODE REVIEW

File: repository/announcement_repository.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+from schemas.announcement_schema import AnnouncementCreate, AnnouncementUpdate
+from utils.singleton_meta import SingletonMeta
+
+
+class AnnouncementRepository(metaclass=SingletonMeta):
+    def __init__(self, session: Session):
+        self.session = session
+
+    def add_announcement(self, announcement: AnnouncementCreate) -> Announcement:
+        # Create a new Announcement object from the AnnouncementCreate schema
```
Comment (Renwarren) [inline repository/announcement_repository.py:13]: Unnecessary comment

File: repository/announcement_repository.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+    def delete_announcement(self, announcement_id: int) -> Optional[Announcement]:
+        db_announcement = self.get_announcement_by_id(announcement_id)
+        if db_announcement:
+            self.session.delete(db_announcement)
+            self.session.commit()
+
+        return db_announcement
+
+    def get_announcement_by_title(self, title) -> Optional[Announcement]:
```
Comment (Renwarren) [inline repository/announcement_repository.py:46]: Why this method?

File: schemas/announcement_schema.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    announcement_deadline: Optional[datetime] = None
+    image: Optional[str] = None
+
+    class Config:
+        orm_mode = True
+        from_attributes = True
+
+
+class AnnouncementUpdate(BaseModel):
+    id: int
```
Comment (Renwarren) [inline schemas/announcement_schema.py:21]: I do not think we should allow people to change the id

File: schemas/announcement_schema.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+class AnnouncementCreate(BaseModel):
+    id: Optional[int] = None
+    title: str
+    description: str
+    announcement_date: datetime
+    announcement_deadline: Optional[datetime] = None
+    image: Optional[str] = None
+
+    class Config:
```
Comment (Renwarren) [inline schemas/announcement_schema.py:15]: this is deprecated. Use <model_config = ConfigDict(from_attributes=True)>

File: services/announcement_service.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+from schemas.announcement_schema import AnnouncementCreate, AnnouncementUpdate, AnnouncementResponse
+
+
+class AnnouncementService:
+
+    def __init__(self, session: Session):
+        self.repository = AnnouncementRepository(session)
+
+    def create_announcement(self, announcement: AnnouncementCreate) -> AnnouncementResponse:
+        # check if announcement already exists
```
Comment (Renwarren) [inline services/announcement_service.py:14]: unncessary comment

File: repository/announcement_repository.py (hunk 2 above)
Comment (atakoutene) [reply repository/announcement_repository.py:46]: I used it to check for duplicates announcement.

File: repository/announcement_repository.py (hunk 1 above)
Comment (atakoutene) [reply repository/announcement_repository.py:13]: Ok! Thank You!

File: schemas/announcement_schema.py (hunk 3 above)
Comment (atakoutene) [reply schemas/announcement_schema.py:21]: I think adding the id over there is a way to say we should have the id of the announcement to be updated! But your point made me add a condition to ignore the id if the client attempts to update it. Thank you!

File: schemas/announcement_schema.py (hunk 4 above)
Comment (atakoutene) [reply schemas/announcement_schema.py:15]: Done! Thanks

File: services/announcement_service.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+            raise ValueError("Announcement date must be before announcement deadline.")
+
+        return AnnouncementResponse.from_orm(self.repository.add_announcement(announcement))
+
+    def get_all_announcements(self) -> list[AnnouncementCreate]:
+        announcements = self.repository.get_announcements()
+        return [AnnouncementCreate.from_orm(announcement) for announcement in announcements]
+
+    def get_announcement_by_id(self, announcement_id: int) -> Optional[AnnouncementCreate]:
+        return AnnouncementCreate.from_orm(self.repository.get_announcement_by_id(announcement_id))
```
Comment (Renwarren) [inline services/announcement_service.py:29]: `from_orm` is deprecated, use:
set `model_config['from_attributes']=True` and use `model_validate` instead.

File: repository/announcement_repository.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+
+class AnnouncementRepository(metaclass=SingletonMeta):
+    def __init__(self, session: Session):
+        self.session = session
+
+    def add_announcement(self, announcement: AnnouncementCreate) -> Announcement:
+        # Create a new Announcement object from the AnnouncementCreate schema
+        db_announcement = Announcement()
+        for key, value in announcement.dict().items():
```
Comment (Renwarren) [inline repository/announcement_repository.py:15]: The `dict` method is deprecated; use `model_dump` instead

File: .idea/.gitignore
Comment (Renwarren) [inline .idea/.gitignore:1]: Mark the .idea folder in the global .gitignore file so it is not pushed

**Review by Renwarren — CHANGES_REQUESTED** [review]

**Comment by Renwarren** [comment]
A test form postman would also look good.

**Comment by Renwarren** [comment]
I think something as News can be something as simple as "Launched V1 of the website", that's a piece of news without a deadline. An announcement might be "NSBE 2025 Scholarship Registration" which would have a a deadline. 

Let me know what you think

**Comment by atakoutene** [comment]
> I think something as News can be something as simple as "Launched V1 of the website", that's a piece of news without a deadline. An announcement might be "NSBE 2025 Scholarship Registration" or "Christmas Party 12/25" which would have a a deadline.
I see what you mean, I think we can treat everything as announcement, since the deadline field is optional, if in the future we need to handle news separately, we'll just follow what have been done for announcement.

**Comment by Renwarren** [comment]
I updated the WHY of your PR to show you what was expected.
Also it seems you have some merge conflict, can you fix them first

File: .gitignore [hunk 8]
```diff
@@ -25,6 +27,13 @@ share/python-wheels/
 .installed.cfg
 *.egg
 MANIFEST
+.idea/.gitignore
```
Comment (Renwarren) [inline .gitignore:30]: you can just do .idea/

File: repository/announcement_repository.py [hunk 9]
```diff
@@ -40,8 +36,7 @@ def delete_announcement(self, announcement_id: int) -> Optional[Announcement]:
         if db_announcement:
             self.session.delete(db_announcement)
             self.session.commit()
-
         return db_announcement
 
-    def get_announcement_by_title(self, title) -> Optional[Announcement]:
-        return self.session.query(Announcement).filter(Announcement.title == title).first()
+    def get_announcement_by_title(self, title: str) -> Optional[Announcement]:
```
Comment (Renwarren) [inline repository/announcement_repository.py:41]: Nice

File: schemas/announcement_schema.py [hunk 10]
```diff
@@ -18,26 +18,20 @@ class Config:
...
 class AnnouncementResponse(BaseModel):
     id: Optional[int] = None
     message: Optional[str] = None
 
-    class Config:
-        orm_mode = True
-        from_attributes = True
 
-    #model_config = ConfigDict(from_attributes=True)
+    model_config = ConfigDict(from_attributes=True, orm_mode=True)
```
Comment (Renwarren) [inline schemas/announcement_schema.py:37]: nice

File: .gitignore (hunk 8 above)
Comment (atakoutene) [reply .gitignore:30]: ok I see.

**Comment by Renwarren** [comment]
LGTM!!!

**Review by Renwarren — APPROVED** [review]

**Pull request merged**